File: src/model/nostr/parser/kind7376.go
// kind7376.go - Spending History Event

package parser

import (
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/candypoets/nutscash/nostr/types"
	"github.com/nbd-wtf/go-nostr"
)

// HistoryTag represents a tag in the spending history
type HistoryTag struct {
	Name   string `json:"name" msgpack:"name"`
	Value  string `json:"value" msgpack:"value"`
	Relay  string `json:"relay,omitempty" msgpack:"relay,omitempty"`
	Marker string `json:"marker,omitempty" msgpack:"marker,omitempty"`
}

// Kind7376Parsed represents a parsed spending history event
type Kind7376Parsed struct {
	Direction       string       `json:"direction" msgpack:"direction"`             // "in" or "out"
	Amount          int          `json:"amount" msgpack:"amount"`                   // Amount in sats
	CreatedEvents   []string     `json:"createdEvents" msgpack:"createdEvents"`     // IDs of token events created
	DestroyedEvents []string     `json:"destroyedEvents" msgpack:"destroyedEvents"` // IDs of token events destroyed
	RedeemedEvents  []string     `json:"redeemedEvents" msgpack:"redeemedEvents"`   // IDs of NIP-61 nutzap events redeemed
	Tags            []HistoryTag `json:"tags,omitempty" msgpack:"tags,omitempty"`   // All decrypted tags
	Decrypted       bool         `json:"decrypted" msgpack:"decrypted"`             // Whether content was successfully decrypted
}

// ParseKind7376 parses a kind 7376 (spending history) event
func (p *Parser) ParseKind7376(event nostr.Event) (*Kind7376Parsed, *[]types.Request, error) {
	if event.Kind != 7376 {
		return nil, nil, fmt.Errorf("event is not kind 7376")
	}

	var requests = []types.Request{}

	parsed := &Kind7376Parsed{
		Decrypted: false,
	}

	// Process unencrypted e tags with "redeemed" marker
	for _, tag := range event.Tags {
		if len(tag) >= 4 && tag[0] == "e" && tag[3] == "redeemed" {
			parsed.RedeemedEvents = append(parsed.RedeemedEvents, tag[1])
			// append a new request
			requests = append(requests, types.Request{
				Kinds:      []int{7375},
				IDs:        []string{tag[1]},
				CacheFirst: true,
				Relays:     p.GetRelays(7375, event.PubKey),
			})
		}
	}

	currentSigner := p.Signer.Current

	// Return early if there's no signer
	if currentSigner == nil {
		return nil, &requests, fmt.Errorf("no signer available for decryption")
	}

	// Try to decrypt the content if signer is available
	pubkey, _ := currentSigner.GetPublicKey()
	decrypted, err := currentSigner.NIP44Decrypt(pubkey, event.Content)
	if err == nil && decrypted != "" {
		var tags [][]string
		if err := json.Unmarshal([]byte(decrypted), &tags); err == nil {
			parsed.Decrypted = true
			parsed.Tags = make([]HistoryTag, 0, len(tags))

			// Process decrypted tags
			for _, tag := range tags {
				if len(tag) >= 2 {
					historyTag := HistoryTag{
						Name:  tag[0],
						Value: tag[1],
					}

					if len(tag) >= 3 {
						historyTag.Relay = tag[2]
					}

					if len(tag) >= 4 {
						historyTag.Marker = tag[3]
					}

					parsed.Tags = append(parsed.Tags, historyTag)

					// Extract specific tag values
					switch tag[0] {
					case "direction":
						parsed.Direction = tag[1]
					case "amount":
						if amt, err := strconv.Atoi(tag[1]); err == nil {
							parsed.Amount = amt
						}
					case "e":
						if len(tag) >= 4 {
							switch tag[3] {
							case "created":
								parsed.CreatedEvents = append(parsed.CreatedEvents, tag[1])
								requests = append(requests, types.Request{
									Kinds:      []int{7375},
									IDs:        []string{tag[1]},
									CacheFirst: true,
									Relays:     p.GetRelays(7375, event.PubKey),
								})
							case "destroyed":
								parsed.DestroyedEvents = append(parsed.DestroyedEvents, tag[1])
								requests = append(requests, types.Request{
									Kinds:      []int{7375},
									IDs:        []string{tag[1]},
									CacheFirst: true,
									Relays:     p.GetRelays(7375, event.PubKey),
								})
							case "redeemed":
								parsed.RedeemedEvents = append(parsed.RedeemedEvents, tag[1])
								requests = append(requests, types.Request{
									Kinds:      []int{7375},
									IDs:        []string{tag[1]},
									CacheFirst: true,
									Relays:     p.GetRelays(7375, event.PubKey),
								})
							}
						}
					}
				}
			}
		}
	}

	return parsed, &requests, nil
}

// PrepareKind7376 prepares a kind 7376 event for publishing by encrypting content and signing
func (p *Parser) PrepareKind7376(event *nostr.Event) error {
	if event.Kind != 7376 {
		return fmt.Errorf("event is not kind 7376")
	}

	// For spending history events, the content is an array of tags
	var tags [][]string
	if err := json.Unmarshal([]byte(event.Content), &tags); err != nil {
		return fmt.Errorf("invalid spending history content: %w", err)
	}

	// Check for required direction and amount tags
	var hasDirection, hasAmount bool
	for _, tag := range tags {
		if len(tag) >= 2 {
			if tag[0] == "direction" {
				hasDirection = true
				if tag[1] != "in" && tag[1] != "out" {
					return fmt.Errorf("direction must be 'in' or 'out'")
				}
			} else if tag[0] == "amount" {
				hasAmount = true
			}
		}
	}

	if !hasDirection || !hasAmount {
		return fmt.Errorf("spending history must include direction and amount")
	}

	// NIP-44 encrypt the content
	tagsJSON, err := json.Marshal(tags)
	if err != nil {
		return fmt.Errorf("failed to marshal tags: %w", err)
	}

	currentSigner := p.Signer.Current

	if currentSigner == nil {
		return fmt.Errorf("no signer available for encryption")
	}

	pubkey, _ := currentSigner.GetPublicKey()

	encrypted, err := currentSigner.NIP44Encrypt(pubkey, string(tagsJSON))
	if err != nil {
		return fmt.Errorf("failed to encrypt tags: %w", err)
	}
	event.Content = encrypted

	// Sign the event
	return currentSigner.SignEvent(event)
}

File: src/model/nostr-worker/src/parser/kind7376.rs
use crate::parser::Parser;
use crate::types::network::Request;
use anyhow::{anyhow, Result};
use nostr::{Event, UnsignedEvent};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryTag {
    pub name: String,
    pub value: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relay: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub marker: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind7376Parsed {
    pub direction: String, // "in" or "out"
    pub amount: i32,       // Amount in sats
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(rename = "createdEvents")]
    pub created_events: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(rename = "destroyedEvents")]
    pub destroyed_events: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(rename = "redeemedEvents")]
    pub redeemed_events: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub tags: Vec<HistoryTag>,
    pub decrypted: bool,
}

impl Parser {
    pub fn parse_kind_7376(&self, event: &Event) -> Result<(Kind7376Parsed, Option<Vec<Request>>)> {
        if event.kind.as_u64() != 7376 {
            return Err(anyhow!("event is not kind 7376"));
        }

        let mut requests = Vec::new();
        let mut parsed = Kind7376Parsed {
            direction: String::new(),
            amount: 0,
            created_events: Vec::new(),
            destroyed_events: Vec::new(),
            redeemed_events: Vec::new(),
            tags: Vec::new(),
            decrypted: false,
        };

        // Process unencrypted e tags with "redeemed" marker
        for tag in &event.tags {
            let tag_vec = tag.as_vec();
            if tag_vec.len() >= 4 && tag_vec[0] == "e" && tag_vec[3] == "redeemed" {
                parsed.redeemed_events.push(tag_vec[1].clone());
                // Add request for this event
                requests.push(Request {
                    ids: vec![tag_vec[1].clone()],
                    kinds: vec![7375],
                    relays: self.database.find_relay_candidates(
                        7375,
                        &event.pubkey.to_hex(),
                        &false,
                    ),
                    cache_first: true,
                    ..Default::default()
                });
            }
        }

        let signer = &self.signer_manager;

        if signer.has_signer() {
            let pubkey = signer.get_public_key()?;
            if let Ok(decrypted) = signer.nip44_decrypt(&pubkey, &event.content) {
                if !decrypted.is_empty() {
                    if let Ok(tags) = serde_json::from_str::<Vec<Vec<String>>>(&decrypted) {
                        parsed.decrypted = true;
                        parsed.tags = Vec::new();

                        // Process decrypted tags
                        for tag in tags {
                            if tag.len() >= 2 {
                                let history_tag = HistoryTag {
                                    name: tag[0].clone(),
                                    value: tag[1].clone(),
                                    relay: tag.get(2).cloned(),
                                    marker: tag.get(3).cloned(),
                                };
                                parsed.tags.push(history_tag);

                                // Extract specific tag values
                                match tag[0].as_str() {
                                    "direction" => parsed.direction = tag[1].clone(),
                                    "amount" => {
                                        if let Ok(amt) = tag[1].parse::<i32>() {
                                            parsed.amount = amt;
                                        }
                                    }
                                    "e" => {
                                        if tag.len() >= 4 {
                                            match tag[3].as_str() {
                                                "created" => {
                                                    parsed.created_events.push(tag[1].clone())
                                                }
                                                "destroyed" => {
                                                    parsed.destroyed_events.push(tag[1].clone())
                                                }
                                                "redeemed" => {
                                                    parsed.redeemed_events.push(tag[1].clone())
                                                }
                                                _ => {}
                                            }
                                        }
                                    }
                                    _ => {}
                                }
                            }
                        }
                    }
                }
            }
        }

        Ok((parsed, Some(requests)))
    }

    pub fn prepare_kind_7376(&self, unsigned_event: &mut UnsignedEvent) -> Result<Event> {
        if unsigned_event.kind.as_u64() != 7376 {
            return Err(anyhow!("event is not kind 7376"));
        }

        // For spending history events, the content is an array of tags
        let tags: Vec<Vec<String>> = serde_json::from_str(&unsigned_event.content)
            .map_err(|e| anyhow!("invalid spending history content: {}", e))?;

        // Check for required direction and amount tags
        let mut has_direction = false;
        let mut has_amount = false;

        for tag in &tags {
            if tag.len() >= 2 {
                match tag[0].as_str() {
                    "direction" => {
                        has_direction = true;
                        if tag[1] != "in" && tag[1] != "out" {
                            return Err(anyhow!("direction must be 'in' or 'out'"));
                        }
                    }
                    "amount" => has_amount = true,
                    _ => {}
                }
            }
        }

        if !has_direction || !has_amount {
            return Err(anyhow!(
                "spending history must include direction and amount"
            ));
        }

        // NIP-44 encrypt the content
        let tags_json =
            serde_json::to_string(&tags).map_err(|e| anyhow!("failed to marshal tags: {}", e))?;

        let signer_manager = &self.signer_manager;

        if !signer_manager.has_signer() {
            return Err(anyhow!("no signer available for encryption"));
        }

        let pubkey = signer_manager.get_public_key()?;

        let encrypted = signer_manager
            .nip44_encrypt(&pubkey, &tags_json)
            .map_err(|e| anyhow!("failed to encrypt tags: {}", e))?;

        unsigned_event.content = encrypted;

        // Sign the event
        signer_manager
            .sign_event(unsigned_event)
            .map_err(|e| anyhow!("failed to sign event: {}", e))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use nostr::{EventBuilder, Keys, Kind, Tag};

    #[test]
    fn test_parse_kind_7376_basic() {
        let keys = Keys::generate();

        let event = EventBuilder::new(Kind::Custom(7376), "encrypted_content", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let (parsed, _) = parser.parse_kind_7376(&event).unwrap();

        assert!(!parsed.decrypted); // No decryption without signer
        assert!(parsed.direction.is_empty());
        assert_eq!(parsed.amount, 0);
    }

    #[test]
    fn test_parse_kind_7376_with_redeemed_tag() {
        let keys = Keys::generate();
        let redeemed_event_id = "1234567890abcdef1234567890abcdef12345678";

        let tags = vec![Tag::parse(vec![
            "e".to_string(),
            redeemed_event_id.to_string(),
            "".to_string(),
            "redeemed".to_string(),
        ])
        .unwrap()];

        let event = EventBuilder::new(Kind::Custom(7376), "encrypted_content", tags)
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let (parsed, requests) = parser.parse_kind_7376(&event).unwrap();

        assert_eq!(parsed.redeemed_events, vec![redeemed_event_id]);
        assert!(requests.is_some());
        assert!(!requests.unwrap().is_empty());
    }

    #[test]
    fn test_prepare_kind_7376_invalid_content() {
        // let keys = Keys::generate();

        // let mut event = EventBuilder::new(Kind::Custom(7376), "invalid json", Vec::new())
        //     .to_event(&keys)
        //     .unwrap();

        // let parser = Parser::default();
        // let result = parser.prepare_kind_7376(&mut event);

        // assert!(result.is_err());
        // assert!(result
        //     .unwrap_err()
        //     .to_string()
        //     .contains("invalid spending history content"));
    }

    #[test]
    fn test_prepare_kind_7376_missing_required_fields() {
        // let keys = Keys::generate();
        // let content = r#"[["amount", "100"]]"#; // Missing direction

        // let mut event = EventBuilder::new(Kind::Custom(7376), content, Vec::new())
        //     .to_event(&keys)
        //     .unwrap();

        // let parser = Parser::default();
        // let result = parser.prepare_kind_7376(&mut event);

        // assert!(result.is_err());
        // assert!(result
        //     .unwrap_err()
        //     .to_string()
        //     .contains("must include direction and amount"));
    }

    #[test]
    fn test_parse_wrong_kind() {
        let keys = Keys::generate();

        let event = EventBuilder::new(Kind::TextNote, "test", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.parse_kind_7376(&event);

        assert!(result.is_err());
    }
}
